### ISSUE TITLE
Regression `used-before-assignment` with type annotation

### ISSUE THREAD

**Issue by cdce8p** [PyCQA/pylint#5342]
### Bug description

```py
from typing import Optional

class A:
    def func(self) -> None:
        _node: Optional[A] = self  # false-positive: used-before-assignment
```

Noticed this in `astroid` first -> [astroid/nodes/node_ng.py](https://github.com/PyCQA/astroid/blob/96ca4dc75aacaba4d1f4e9a66a72699c9fc93349/astroid/nodes/node_ng.py#L435). The example above is valid as the function body isn't evaluated until called at witch time the class is already defined. This is a regression in `main` compared to `v2.11.1`.

/CC: @DanielNoord Haven't looked at it yet, but I would guess this might be related to the work you've been doing in this area.

### Command used

```shell
pylint test.py
```

### Pylint output

```shell
E0601: Using variable 'A' before assignment (used-before-assignment)
```

### Expected behavior

No error

### Pylint version

```shell
pylint 2.11.2-dev0
astroid 2.8.6-dev0
Python 3.10.0 (v3.10.0:b494f5935c, Oct  4 2021, 14:59:20) [Clang 12.0.5 (clang-1205.0.22.11)]
```

**Comment by DanielNoord** [comment]
I'll have a look!

**Comment by DanielNoord** [comment]
For future reference.

This does create a false positive:

```python
class MyClass:
    def self_referential_optional_within_method(self) -> None:
        variable: Optional[MyClass] = self
        print(variable)

    def correct_inner_typing_method(self) -> bool:
        def inner_method(self, other: MyClass) -> bool:
            return self == other

        return inner_method(self, MyClass())
```

But this doesn't:

```python
class MyClass:
    def correct_inner_typing_method(self) -> bool:
        def inner_method(self, other: MyClass) -> bool:
            return self == other

        return inner_method(self, MyClass())
    
    def self_referential_optional_within_method(self) -> None:
        variable: Optional[MyClass] = self
        print(variable)
```

**Comment by The-Compiler** [comment]
FWIW I can confirm this seems fully fixed in 2.12.2 - thanks everyone!
